src/types.ts

```diff
@@ -0,0 +1,128 @@
+} from "@humanwhocodes/momoa";
+
+//------------------------------------------------------------------------------
+// Types
+//------------------------------------------------------------------------------
+
+type ValueNodeParent = DocumentNode | MemberNode | ElementNode;
+
+/**
+ * A JSON syntax element, including nodes and tokens.
+ */
+export type JSONSyntaxElement = Token | AnyNode;
+
+/**
+ * Language options provided for JSON files.
+ */
+export interface JSONLanguageOptions extends LanguageOptions {
+	/**
+	 * Whether to allow trailing commas. Only valid in JSONC.
+	 */
+	allowTrailingCommas?: boolean;
+}
+
+/**
+ * The visitor format returned from rules in this package.
+ */
+export interface JSONRuleVisitor extends RuleVisitor {
+	Document?(node: DocumentNode): void;
+	Member?(node: MemberNode, parent?: ObjectNode): void;
+	Element?(node: ElementNode, parent?: ArrayNode): void;
+	Object?(node: ObjectNode, parent?: ValueNodeParent): void;
+	Array?(node: ArrayNode, parent?: ValueNodeParent): void;
+	String?(node: StringNode, parent?: ValueNodeParent): void;
+	Null?(node: NullNode, parent?: ValueNodeParent): void;
+	Number?(node: NumberNode, parent?: ValueNodeParent): void;
+	Boolean?(node: BooleanNode, parent?: ValueNodeParent): void;
+	NaN?(node: NaNNode, parent?: ValueNodeParent): void;
+	Infinity?(node: InfinityNode, parent?: ValueNodeParent): void;
+	Identifier?(node: IdentifierNode, parent?: ValueNodeParent): void;
+
+	"Document:exit"?(node: DocumentNode): void;
+	"Member:exit"?(node: MemberNode, parent?: ObjectNode): void;
+	"Element:exit"?(node: ElementNode, parent?: ArrayNode): void;
+	"Object:exit"?(node: ObjectNode, parent?: ValueNodeParent): void;
+	"Array:exit"?(node: ArrayNode, parent?: ValueNodeParent): void;
+	"String:exit"?(node: StringNode, parent?: ValueNodeParent): void;
+	"Null:exit"?(node: NullNode, parent?: ValueNodeParent): void;
+	"Number:exit"?(node: NumberNode, parent?: ValueNodeParent): void;
+	"Boolean:exit"?(node: BooleanNode, parent?: ValueNodeParent): void;
+	"NaN:exit"?(node: NaNNode, parent?: ValueNodeParent): void;
+	"Infinity:exit"?(node: InfinityNode, parent?: ValueNodeParent): void;
+	"Identifier:exit"?(node: IdentifierNode, parent?: ValueNodeParent): void;
+}
+
+/**
+ * The `SourceCode` implementation for JSON files.
+ */
+export interface IJSONSourceCode
+	extends TextSourceCode<{
+		LangOptions: JSONLanguageOptions;
+		RootNode: DocumentNode;
+		SyntaxElementWithLoc: JSONSyntaxElement;
+		ConfigNode: Token;
+	}> {
+	/**
+	 * Get the text of a syntax element.
+	 * @param syntaxElement The syntax element to get the text of.
+	 * @param beforeCount The number of characters to include before the syntax element.
+	 * @param afterCount The number of characters to include after the syntax element.
+	 * @returns The text of the syntax element.
+	 */
+	getText(
+		syntaxElement: JSONSyntaxElement,
+		beforeCount?: number,
+		afterCount?: number,
+	): string;
+}
+
+export type IJSONLanguage = Language<{
```

Comment:
[Style] Was the `I` an intentional addition? It's not typical in TypeScript code.

```suggestion
export type JSONLanguage = Language<{
```

Here and with `IJSONSourceCode`.

Reply:
Yes, because there's already a class called `JSONLanguage`, so the type is `IJSONLanguage`. The `I` stands for interface.

I have seen this in other projects, and it's fairly typical in languages like Java where you can define an interface separate from a class.

Reply:
Quick answer: there isn't a single standard, but a name like `JSONLanguageLike` would be reasonable.

---

Longer answer: the _"interface only used for one class"_ pattern isn't common in TypeScript-land. If the interface is really only used for that one class then most of the time folks would just use the class name.

Since we're not beholden to classic Java-style class hierarchies, it's less common AFAIU to put everything in a single shape the way the `@eslint/core` `Language` class is set up. A lot of architectures avoid a specific classes altogether and instead just go with generic factory functions. 

Quickly sketching a theoretical vague equivalent:

```ts
declare function createLanguage<Settings extends LanguageSettings>
  (settings: Settings): Language<Settings>;

export const jsonLanguage = createLanguage({
  fileType: "text",
  lineStart: 1,
  // ...
});
```

In that world, there wouldn't be a need for a `JSONLanguage` to be explicitly declared. It could be inferred from `typeof jsonLanguage` with a helper.

Not suggesting changes, just posting context for why this naming problem isn't as commonly solved.